### ISSUE TITLE
generated canvas has doubled dimensions

### ISSUE THREAD

**Issue by Vaguery** [quil/quil#299]
I'm trying to track down a bug when embedding a `quil` sketch in an external page, and I've noticed that when I specify a `q/sketch` parameter of `:size [500 500]`, I see the following generated canvas inside the target `div` on the sketch page:

```
<canvas id="defaultCanvas0" class="p5Canvas" width="1000" height="1000" style="width: 500px; height: 500px;"></canvas>
```

I'd really rather have the `width` and `height` match the intended values.

**Comment by nbeloglazov** [comment]
@anthonygalea I assume width/height is generated by pj5s? might be a bug there.

**Comment by anthonygalea** [comment]
Yes @nbeloglazov just reproduced this with a p5js sketch using both `v0.7.3` (the version quil is currently using) as well as the latest p5js release `v0.8.0`. You can also see it in the [p5js editor](https://editor.p5js.org/). Did a quick search and I don't see a bug already reported. @Vaguery can you report this on the [p5js issue tracker](https://github.com/processing/p5.js/issues)? Once it's fixed there we can update the dependency in quil.

**Comment by anthonygalea** [comment]
https://github.com/processing/p5.js/blob/master/src/core/p5.Renderer.js#L103
@Vaguery are you using a hidpi display?

**Comment by Vaguery** [comment]
@anthonygalea Yes (it's all I have, in fact): MBP, iMac, iPhone, iPad.

**Comment by Vaguery** [comment]
I suppose then that this is... intentional? Is this an anti-aliasing behavior in `p5.js`?

**Comment by anthonygalea** [comment]
Not sure but we'll find out: https://discourse.processing.org/t/should-width-and-height-of-canvas-be-double/10451

**Comment by Vaguery** [comment]
Yes, this whole pile of advice seems salient:

https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/HTML-canvas-guide/SettingUptheCanvas/SettingUptheCanvas.html

**Comment by Vaguery** [comment]
I think it probably all _is_ an intentional design decision (and not a bad one, given the hardware ecosystem out there).

The specific bug that motivated my check involved the display of a supposedly "400px" square `quil` sketch on a HiDPI iOS device in portrait mode &c &c &c... it gets pretty specific.

The reasonable-feeling fix, for me, is to work with the CSS on the pages where such `quil` sketches are embedded. So here's the approach I'm using now, and it seems to have fixed the issue, for me, and _for the moment_:

``` css
.quil-sketch {
  height: 0;
  width: 100%;
  padding-bottom: 100%;
}

canvas {
  max-width: 100%;
  max-height: 100%;
}
```

That is, I specified the `div` host style to "want to be square", and forced all `canvas` sizes to have maxima as indicated. It feels like there will be a problem some day farther on, maybe as soon as I want to have a non-square `quil` sketch, but for now this handles my problem on iOS with overflow and scaling.

Perhaps though the "real bug" is that this should probably be better documented, since it seems to be (1) a good thing to have pretty graphics on expensive hardware, but (2) a nasty thing to try to debug if you don't notice the `width` tricks in `p5.js`....

**Comment by Vaguery** [comment]
☝️ actually, does this feel like the sort of thing that would work as a `quil` pull request? I imagine that anybody who embeds any `quil` sketch in a web page with a plain `div` will experience a CSS problem in certain overflow conditions.

**Comment by nbeloglazov** [comment]
Bill, can you provide an example (with full html) which is not working as expected? I feel like inline width/height CSS style should overwrite width/height attributes so I don't fully understand what's broken.

**Comment by professorwright** [comment]
I'm not exactly sure what the status is here, but I think this issue might be affecting me. In my case, I am trying to make a cljs sketch which is responsive to window resize events. When I call `quil.core/resize`, strange behavior ensues. One specific observation I can make is that the relationship between CSS width/height and canvas width/height is lost. While initially the canvas will have a width and height of, say, 1400x1000, double that of the CSS height and width of 700x500, after calling `quil.core/resize`, all widths and heights are equal. The 2:1 relationship is not maintained. Please advise.

**Comment by nbeloglazov** [comment]
I see. Then it is a bug in `resize-sketch` function, it doesn't account for the screen resolution [here](https://github.com/quil/quil/blob/master/src/cljs/quil/sketch.cljs#L23-L35) as it should. Easiest solution for you would be just copy that function locally and use [display-density](http://quil.info/api/environment#display-density) to set width/height. Meanwhile we need to fix it in Quil.

**Comment by Hodapp87** [comment]
Running into the same thing here in ClojureScript on Linux (on a 4K display), including with sketches run through quil.info. `update-pixels` and `aset` are pretty much broken.

**Comment by thinsoldier** [comment]
Wow! Does this problem STILL exist in 2021?

**Comment by meosr** [comment]
Had the same problem right now. Everything was fixed by setting pixelDensity(1) in my sketch setup. 
It was a problem for me, because I wanted to post the canvas to a server automatically and on my Ubuntu the file was sent with correct size, and on Windows the size was doubled, which then prevented me from displaying saved canvases correctly.

**Comment by JosefBredereck** [comment]
Facing the same issue in 2022 🆘

**Comment by dgtized** [comment]
I *think* this was resolved in the 4.3.1323 release. Can someone else verify?
